### ISSUE TITLE
graphical-session.target does not start when running wayland sessions

### ISSUE THREAD

**Issue by nrdxp** [NixOS/nixpkgs#169143]
### Describe the bug
The graphical-session.target systemd user target doesn't seem to get started when running a session from a display manager (in my case GDM), if that session is wayland based. Looks like the reason is that this is usually started from the [xsession wrapper script](https://github.com/NixOS/nixpkgs/blob/fd609f92336b4c1677697b09202cfc1439f35aec/nixos/modules/services/x11/display-managers/default.nix#L90), which obviously doesn't play a role in wayland sessions.

### Steps To Reproduce
Easiest way to reproduce might be enable GDM and Sway and start a session. Then just confirm with `systemctl --user status graphical-session.target`.

### Expected behavior
Should work the same way as in an X session, that is `graphical-session.target` is started on a new session and stopped on session exit, allow user units which are wanted by the graphical-session to also start and stop as expected on session starts and restarts.

**Comment by peterhoeg** [comment]
With plasma launched from sddm, `graphical-session.target` *is* being started:

```
systemctl --user status graphical-session.target
● graphical-session.target - Current graphical user session
     Loaded: loaded (/etc/systemd/user/graphical-session.target; static)
    Drop-In: /nix/store/3nkqvbc3di65kgjprwbsbaffm0irphsx-user-units/graphical-session.target.d
             └─overrides.conf
     Active: active since Tue 2022-04-19 19:19:14 +08; 56min ago
      Until: Tue 2022-04-19 19:19:14 +08; 56min ago
       Docs: man:systemd.special(7)

Apr 19 19:19:14 dolores systemd[1021]: Reached target Current graphical user session.
```

**Comment by bandithedoge** [comment]
This also happens with X sessions started with sx.

**Comment by kira-bruneau** [comment]
Depending on the login manager you use, this won't be started if it doesn't use [`xsessionWrapper`](https://github.com/NixOS/nixpkgs/blob/4a1f5a1a7da8bcf205f4f691408f8cc0bd906fee/nixos/modules/services/x11/display-managers/default.nix#L90) & [`xsession`](https://github.com/NixOS/nixpkgs/blob/4a1f5a1a7da8bcf205f4f691408f8cc0bd906fee/nixos/modules/services/x11/display-managers/default.nix#L450).

To workaround this with greetd, I created a `wayland-session` wrapper that handles this: https://gitlab.com/kira-bruneau/nixos-config/-/blob/main/environment/desktop.nix#L137-148.

I was planning on upstreaming my changes, but I wasn't quite sure how to generalize it, and I really wanted to look into refactoring the messiness between `xsession` & `xsessionWrapper` (I think it should just be one wrapper script).

**Comment by nrdxp** [comment]
Maybe the reason it doesn't work with GDM on my system is because GDM is using the wayland backend by default and so no X is ever started. I don't know exactly I only traced this to the xsession wrapper, but didn't review what display managers call this unconditionally and which don't. I'll try to modify the title of the issue to more accurately reflect the current situation.

**Comment by kira-bruneau** [comment]
Yeah, it looks like `GDM_X_SESSION_WRAPPER` is only ever used when running X, but it still calls xsession indirectly through the `/share/xsessions/<session-name>.desktop` files.

**Comment by nrdxp** [comment]
Yeah looks like the xsession script is simply appened to all the files in the xsession directory ending in `-xsession`, so perhaps, as a minimal fix we just need to modify the following lines to do the same for files in the `wayland-sessions` directory?
https://github.com/danme/nixpkgs/blob/bd9c5b933cfe4b3bc71b962fe6436f5d373c3b29/nixos/modules/services/x11/display-managers/default.nix#L418-L419

I think modify the script to eventually make it more general as you recommended would be a better long term solution though.

**Comment by kira-bruneau** [comment]
I don't think it would actually be that simple, because the xsessionWrapper is what actually starts the target, but it's completely unrelated to the .desktop session files. That's kind of why I was thinking of a way to move everything into `xsession` :sweat_smile:

**Comment by kira-bruneau** [comment]
I'm just trying to wrap my head around all the different parts :sweat_smile: , it definitely seems like it's over complicated and there should be a cleaner way to do this that's easier to reason about.

**Comment by tbaumann** [comment]
I also notice that this behavior when I switched to greedt.
A generic session wrapper seems like a good idea to me.
Perhaps for all session providers registered via `providedSessions`.

I see the sway package makes an attempt to solve it. It adds an exec command to start the `sway-session.target` that `BindsTo=graphical-session.target`

 `sway-session.target` gets activated and with it `graphical-session.target`. But what never happens is to stop the session.
 
 I thought about implementing something like this in the sway wrapper. https://github.com/alebastr/sway-systemd/blob/main/src/session.sh#L124
 But that is in the end quite sway specific.
 
 GDM and Xsession do a much better job at this unfortunately. We regressed hard here.
GDM even (somehow) supports multiple sessions at the same time.

**Comment by tbaumann** [comment]
https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/display-managers/default.nix evolved massively since. None of those improvements are available to wayland sessions.

**Comment by JohnRTitor** [comment]
@jtojnar @fufexan I am noticing this on GDM + Hyprland as well, due to `graphical-session.target` not being started at all in Wayland sessions in GDM, [many services]*https://github.com/search?q=repo%3ANixOS%2Fnixpkgs%20graphical-session.target&type=code) that depend on it do not get started at all, including [polkit-gnome](https://github.com/rxyhn/yuki/blob/2fcbd0c1cde5fcd6e2236b0aa90c72629dbb3740/modules/system/services/gnome/polkit-gnome.nix#L5).

**Comment by midirhee12** [comment]
I guess this means that each of the wayland compositors should start `graphical-session.target` themselves.

**Comment by tbaumann** [comment]
Not explicitly, that would be madness. (Relying on so many places to do the  same and do it right. In fact sway homemanmager starts the `graphical-session.target` but doesn't stop it at the end.)

In my opinion any wayland-session command should have have a generic wayland session wrapper that handles that. Literally the same way xsession is already handled.

I see no reason why that shouldn't be best. I should perhaps try to work on that. But my Nix foo is still weak and I don't have time...

**Comment by midirhee12** [comment]
Mmm, maybe I wasn't clear. I meant through something like `sway-session.target` for each compositor. I think what I said could be misinterpreted as someone needing to start the `graphical-session.target` in their corresponding compositor's config which is not what I was referring to if so. I agree that that would be chaotic.

**Comment by midirhee12** [comment]
I just tested this on SDDM and the issue persists. So maybe the issue is not dependent on the display manager after all or the issue now continues on SDDM as well.

**Comment by JohnRTitor** [comment]
There's a quick fix for this. You can launch your WMs in a wrapper called [UWSM](https://github.com/Vladimir-csp/uwsm).

This makes sure to start `graphical-session-pre.target` and `graphical-session.target`, and imports the necessary variables to systemd.

There's a ongoing PR for this option to be added on Nixpkgs.

https://github.com/NixOS/nixpkgs/pull/335625

You can test the PR by importing the module [like this](https://github.com/NixOS/nixpkgs/pull/335625#issuecomment-2298885767) or just copy paste the module and import in your NixOS configuration.

You must configure your Window Manager appropriately, https://github.com/NixOS/nixpkgs/pull/335625#issue-2472010546

**Comment by midirhee12** [comment]
Can this be closed as uwsm is now merged?

**Comment by tbaumann** [comment]
uwsm integration seems a bit incomplete. But it seems like a viable workaround. (I couldn't test on unstable yet)

**Comment by midirhee12** [comment]
What do you mean by incomplete? As in not merged yet? If so, it was merged 2 days ago.
